psl/parser-database/src/walkers.rs

```diff
@@ -66,7 +66,8 @@ pub(crate) fn newline(source: &str, span: Span) -> NewlineType {
 }
 
 impl crate::ParserDatabase {
-    fn iter_tops(&self) -> impl Iterator<Item = (FileId, ast::TopId, &ast::Top)> + '_ {
+    /// Iterate all top level blocks.
+    pub fn iter_tops(&self) -> impl Iterator<Item = (FileId, ast::TopId, &ast::Top)> + '_ {
```

Comment:
Current walkers implementations are entirely centred around specific forms of tops, i.e.: `ModelWalker` (this includes views), `EnumWalker`, `CompositeTypeWalker`, and also further walkers of more specific entities.

I'm wondering if it makes sense to include a `TopWalker`.

We can't currently implement a `find_top` that uses the string interner but also gives us access to the `Top` itself

---

When trying to explore implementing a `TopWalker`, I run into the following issue:

```
the type `SchemaAst` cannot be indexed by `(diagnostics::FileId, schema_ast::ast::TopId)`
the trait `Index<(diagnostics::FileId, schema_ast::ast::TopId)>` is not implemented for `SchemaAst`, which is required by `Files: Index<(diagnostics::FileId, (diagnostics::FileId, schema_ast::ast::TopId))>`
the following other types implement trait `Index<Idx>`:
  <SchemaAst as Index<schema_ast::ast::AttributeContainer>>
  <SchemaAst as Index<schema_ast::ast::AttributeId>>
  <SchemaAst as Index<schema_ast::ast::CompositeTypeId>>
  <SchemaAst as Index<schema_ast::ast::ModelId>>
  <SchemaAst as Index<schema_ast::ast::EnumId>>
  <SchemaAst as Index<SourceId>>
  <SchemaAst as Index<schema_ast::ast::TopId>>
```

on the following fn:
```rs
pub type TopWalker<'db> = super::Walker<'db, (FileId, crate::TopId)>;

impl<'db> TopWalker<'db> {
		...
		/// The AST node.
	  pub fn ast_top(self) -> &'db ast::Top {
	      &self.db.asts[self.id]
	  }
}    
```

I note that we have the following type def
```rs
// parser-database/src/ids.rs

/// See [ast::TopId]
pub type TopId = InFile<ast::TopId>;
```

and the following `Index` definition
```rs
// parser-database/src/files.rs

impl<Id> Index<crate::InFile<Id>> for Files
where
    ast::SchemaAst: Index<Id>,
{
    type Output = <ast::SchemaAst as Index<Id>>::Output;

    fn index(&self, index: crate::InFile<Id>) -> &Self::Output {
        &self[index.0].2[index.1]
    }
}
```

But I'm not sure what's missing here

Reply:
You’re using `crate::TopId` instead of `ast::TopId`. `crate::TopId` is an alias for InFile<ast::TopId>, which means that your id field is `(FileId, (FileId, ast::TopId))` instead of `(FileId, ast::Topid)`

```rust
pub type TopWalker<'db> = super::Walker<'db, (FileId, ast::TopId)>;

impl<'db> TopWalker<'db> {
		/// The AST node.
	  pub fn ast_top(self) -> &'db ast::Top {
	      &self.db.asts[self.id]
	  }
} 
```